department-of-veterans-affairs/caseflow
# CAVC extension request action

Bumps caseflow-112

Part 1 of stack
Part 2: #15694

### Description
Add a front end modal to handle extension requests. `onSubmit` need to be fleshed out to actually do something on the back end.

### Acceptance Criteria
- [ ] New modal to allow user to grant/deny extension requests
  - [ ] Selecting "grant" shows on hold duration (and selecting deny hides it)
  - [ ] Selecting "custom" duration shows place to enter a custom duration (and selecting deny or any other duration hides it)
  - [ ] All fields that are currently displayed are required on submit
  - [ ] Completion of grant creates a timed hold task for that long and creates a record keeping CavcGrantedExtensionRequestTask
  - [ ] Completion of deny creates a record keeping CavcDeniedExtensionRequestTask

### Testing Plan
1. run `make storybook` and go to http://localhost:6006/?path=/story/queue-components-cavcreviewextensionrequestmodal--default
2. Log in as CAVC_LIT_SUPPORT_ADMIN and go to http://localhost:3000/organizations/cavc-lit-support
3. Open a Send CAVC Letter task, assign it to yourself, complete it, and end the hold on the new CavcRemandProcessedLetterResponseWindowTask
4. Select "Extension request", Deny, and add instructions.
5. Confirm success message, deny extension appears in case timeline, and you can still act on the task
```ruby
                                                            ┌──────────────────────────────────────────────────────────────────────────────────────────────┐
Appeal 930 (E 201213-929 Court Remand) ──────────────────── │ ID   │ STATUS    │ ASGN_BY                │ ASGN_TO                │ UPDATED_AT              │
└── RootTask                                                │ 2793 │ on_hold   │                        │ Bva                    │ 2020-12-14 17:47:28 UTC │
    └── DistributionTask                                    │ 2794 │ on_hold   │                        │ Bva                    │ 2020-12-14 17:47:28 UTC │
        └── CavcTask                                        │ 2795 │ on_hold   │                        │ Bva                    │ 2020-12-14 17:47:28 UTC │
            ├── SendCavcRemandProcessedLetterTask           │ 2796 │ completed │                        │ CavcLitigationSupport  │ 2020-12-17 21:06:59 UTC │
            │   └── SendCavcRemandProcessedLetterTask       │ 5818 │ completed │ CAVC_LIT_SUPPORT_ADMIN │ CAVC_LIT_SUPPORT_ADMIN │ 2020-12-17 21:06:59 UTC │
            └── CavcRemandProcessedLetterResponseWindowTask │ 5819 │ assigned  │                        │ CavcLitigationSupport  │ 2020-12-17 21:07:12 UTC │
                ├── TimedHoldTask                           │ 5820 │ cancelled │                        │ CavcLitigationSupport  │ 2020-12-17 21:07:04 UTC │
                └── CavcDeniedExtensionRequestTask          │ 5821 │ completed │ CAVC_LIT_SUPPORT_ADMIN │ CAVC_LIT_SUPPORT_ADMIN │ 2020-12-17 21:07:12 UTC │
                                                            └──────────────────────────────────────────────────────────────────────────────────────────────┘
```
6. 4. Select "Extension request", Grant, select a number of days, and add instructions.
5. Confirm success message, grant extension appears in case timeline, and the task is on hold

### User Facing Changes
![Screen Shot 2020-12-15 at 3 31 06 PM](https://user-images.githubusercontent.com/45575454/102283686-e6d1ef80-3f00-11eb-8456-c70cacad9447.png)
```ruby
                                                            ┌──────────────────────────────────────────────────────────────────────────────────────────────┐
Appeal 930 (E 201213-929 Court Remand) ──────────────────── │ ID   │ STATUS    │ ASGN_BY                │ ASGN_TO                │ UPDATED_AT              │
└── RootTask                                                │ 2793 │ on_hold   │                        │ Bva                    │ 2020-12-14 17:47:28 UTC │
    └── DistributionTask                                    │ 2794 │ on_hold   │                        │ Bva                    │ 2020-12-14 17:47:28 UTC │
        └── CavcTask                                        │ 2795 │ on_hold   │                        │ Bva                    │ 2020-12-14 17:47:28 UTC │
            ├── SendCavcRemandProcessedLetterTask           │ 2796 │ completed │                        │ CavcLitigationSupport  │ 2020-12-17 21:06:59 UTC │
            │   └── SendCavcRemandProcessedLetterTask       │ 5818 │ completed │ CAVC_LIT_SUPPORT_ADMIN │ CAVC_LIT_SUPPORT_ADMIN │ 2020-12-17 21:06:59 UTC │
            └── CavcRemandProcessedLetterResponseWindowTask │ 5819 │ on_hold   │                        │ CavcLitigationSupport  │ 2020-12-17 21:07:55 UTC │
                ├── TimedHoldTask                           │ 5820 │ cancelled │                        │ CavcLitigationSupport  │ 2020-12-17 21:07:04 UTC │
                ├── CavcGrantedExtensionRequestTask         │ 5822 │ completed │ CAVC_LIT_SUPPORT_ADMIN │ CAVC_LIT_SUPPORT_ADMIN │ 2020-12-17 21:07:55 UTC │
                └── TimedHoldTask                           │ 5823 │ assigned  │ CAVC_LIT_SUPPORT_ADMIN │ CavcLitigationSupport  │ 2020-12-17 21:07:55 UTC │
                                                            └──────────────────────────────────────────────────────────────────────────────────────────────┘
```


 step|shot
 ---|---
Base Modal|![Screen Shot 2020-12-15 at 3 42 52 PM](https://user-images.githubusercontent.com/45575454/102283739-fe10dd00-3f00-11eb-83f9-b3b8003d52f1.png)
Dynamically show duration dropdown|![Screen Shot 2020-12-15 at 3 43 02 PM](https://user-images.githubusercontent.com/45575454/102283740-fe10dd00-3f00-11eb-9ebf-12a9aa7aa511.png)
Dynamically show custom duration dropdown|![Screen Shot 2020-12-15 at 3 53 01 PM](https://user-images.githubusercontent.com/45575454/102283744-fea97380-3f00-11eb-95a9-136af0b5dcc3.png)
Dynamically hide duration dropdown|![Screen Shot 2020-12-15 at 3 43 07 PM](https://user-images.githubusercontent.com/45575454/102283741-fe10dd00-3f00-11eb-8fe2-75703429bfb0.png)
Decision and instruction validation|![Screen Shot 2020-12-15 at 3 31 29 PM](https://user-images.githubusercontent.com/45575454/102283737-fd784680-3f00-11eb-8d73-99ec1b078e82.png)
Hold duration validation|![Screen Shot 2020-12-15 at 3 31 43 PM](https://user-images.githubusercontent.com/45575454/102283738-fd784680-3f00-11eb-940a-cde83889f9ab.png)
Custom hold duration validatino|![Screen Shot 2020-12-15 at 4 06 26 PM](https://user-images.githubusercontent.com/45575454/102283748-ff420a00-3f00-11eb-8755-f1611fad6ece.png)
Grant success|![Screen Shot 2020-12-17 at 3 49 37 PM](https://user-images.githubusercontent.com/45575454/102543213-f710d880-4080-11eb-864b-f40c52337437.png)
Grant case timeline|![Screen Shot 2020-12-17 at 3 49 46 PM](https://user-images.githubusercontent.com/45575454/102543245-055ef480-4081-11eb-8271-4a90bc5431c7.png)
Deny success|![Screen Shot 2020-12-17 at 3 48 22 PM](https://user-images.githubusercontent.com/45575454/102543195-f24c2480-4080-11eb-827f-d643761e2713.png)
Deny case timeline|![Screen Shot 2020-12-17 at 3 49 07 PM](https://user-images.githubusercontent.com/45575454/102543207-f4ae7e80-4080-11eb-9282-e642ed987f5f.png)


### Code Documentation Updates

### Storybook Story
*For Frontend (Presentationa) Components*
